### PR TITLE
feat(test): print pending tests on sigint

### LINKED ISSUE #17788: Disabling test sanitizers causes test steps to not be logged in time.
Whenever a test gets stuck, it's very difficult to figure out at which step it keeps hanging.

Here's a reduced test case:
```js
import {serve} from "https://deno.land/std@0.177.0/http/mod.ts";

// Start a server to keep the process running without
// 'Promise resolution is still pending but the event loop has already resolved.'
serve(() => new Response("hello"));

Deno.test({
	name: "hanging step",
	sanitizeOps: false,
	sanitizeExit: false,
	sanitizeResources: false,
	async fn(ctx) {
		await ctx.step("Step 1", async (ctx) => {
			await ctx.step("Step 2", async (ctx) => {
				// A promise that never resolves
				await new Promise(r => {});
			});
		});
	}
});
```

Then run with `deno test --allow-net the.test.js`.

For me this keeps hanging without ever logging the steps. Enabling either one of the three disabled sanitizers fixes the issue.

I've been getting a flaky e2e test that I suspect is hard to debug because of this. Though for some reason in that case the name of the test itself isn't being logged either. I think my workaround for now will be to replace all test steps with `console.log`.

### PR DESCRIPTION
**Pull request by nayeemrmn**
Moves test validation/sanitization error formatting to Rust. Moves the
buffer system for reporting concurrent test steps to Rust.

Output is improved:
- Sanitization and other non-user-thrown errors no longer have loud JS
  error stacks.
- Test step failures are now included in the summary.
- Test failures that defer to other test failures (e.g.
  `error: 1 test step failed`) are no longer included in the summary.
  Summaries only contain meaningful errors.
- Fix quirks and inconsistencies caused by the old buffer system.

---

Closes #17788.
```ts
setInterval(() => {}, 10000);

Deno.test({
  name: "test",
  sanitizeOps: false,
  sanitizeExit: false,
  sanitizeResources: false,
  async fn(t) {
    await t.step("step 1", async (t) => {
      await t.step("step 2", async () => {
        await new Promise(() => {});
      });
    });
  },
});
```
image

The vast majority of this PR is the prerequisite of refactoring the concurrent-test-step buffer system to Rust. The buffering logic should be easier to reason about. Instead of changing output streaming behaviour based on sanitizers, it just buffers any test result whose output doesn't belong in the current scope and flushes it when a test on the same scope has finished reporting.

### CODE REVIEW

File: cli/js/40_testing.js [hunk 1]
```diff
@@ -503,17 +503,6 @@ function testStepPostValidation(desc) {
...
-  // check if an ancestor already completed
-  let currentDesc = desc.parent;
-  while (currentDesc != null) {
-    if (MapPrototypeGet(testStates, currentDesc.id).finalized) {
-      throw new Error(
-        "Parent scope completed before test step finished execution. Ensure all steps are awaited (ex. `await t.step(...)`).",
-      );
-    }
-    currentDesc = currentDesc.parent;
-  }
```
Comment (dsherret) [inline cli/js/40_testing.js:516]: Why remove this?

File: cli/tests/testdata/test/steps/invalid_usage.ts [hunk 2]
```diff
@@ -11,7 +11,7 @@ Deno.test("capturing", async (t) => {
 
 Deno.test("top level missing await", (t) => {
   t.step("step", () => {
-    return new Promise((resolve) => setTimeout(resolve, 10));
```
Comment (dsherret) [inline cli/tests/testdata/test/steps/invalid_usage.ts:14]: Why was this changed?

File: cli/tests/testdata/test/steps/invalid_usage.out [hunk 3]
```diff
@@ -9,9 +9,6 @@ top level missing await ... FAILED ([WILDCARD])
 inner missing await ...
   step ...
     inner ... pending ([WILDCARD])
-      error: Error: Parent scope completed before test step finished execution. Ensure all steps are awaited (ex. `await t.step(...)`).
-          at [WILDCARD]
-          at async TestContext.step [WILDCARD]
```
Comment (dsherret) [inline cli/tests/testdata/test/steps/invalid_usage.out:14]: Why was this removed?

Comment (nayeemrmn) [reply cli/tests/testdata/test/steps/invalid_usage.out:14]: This is a redundant error, the parent's error combined with `... pending` represents this. It appears inconsistently as well. For example, turning sanitizers on for `inner missing await` strangely gets rid of it. And it's only reported if the root test scope happens to be open still. Combine that with it creating the need for the error value contained in `TestStepResult::Pending()`, no reason to have it.

File: cli/tests/testdata/test/steps/invalid_usage.ts (hunk 2 above)
Comment (nayeemrmn) [reply cli/tests/testdata/test/steps/invalid_usage.ts:14]: It's an async op leak that's dependent on the following test having sanitizers off. (Some approach I tried while WIP exposed that but it's back to normal)

File: cli/js/40_testing.js (hunk 1 above)
Comment (nayeemrmn) [reply cli/js/40_testing.js:516]: https://github.com/denoland/deno/pull/18246#discussion_r1140919178

File: cli/tests/testdata/test/steps/invalid_usage.out (hunk 3 above)
Comment (dsherret) [reply cli/tests/testdata/test/steps/invalid_usage.out:14]: `pending` isn't very instructive for users though. Can we output an instructive message still?

> It appears inconsistently as well.

We should fix that bug in that case.

Comment (nayeemrmn) [reply cli/tests/testdata/test/steps/invalid_usage.out:14]: > Can we output an instructive message still?

Yes, when a `pending` test step result is printed in rust we can follow that up with any message along with that step's location.

I also wanted to follow up with a change that moved all validation errors to structured error variants which are formatted in Rust instead. We can print a nicely formatted location next to them rather than those loud JS error stacks. It's probably better to do it all at once in this PR.

Comment (nayeemrmn) [reply cli/tests/testdata/test/steps/invalid_usage.out:14]: Done 
image

**Review by bartlomieju — COMMENTED** [review]
I really like how a lot of the logic has been moved from JS to Rust. I have one concern related to signal though - once we install listener for `ctrl+c` then even if it's dropped the process won't exit anymore on SIGINT. 

Is there a chance this can happen with this change? Also how will it interact if the code in tests registers a signal listener for SIGINT event?

**Comment by nayeemrmn** [comment]
> once we install listener for `ctrl+c` then even if it's dropped the process won't exit anymore on SIGINT.
> 
> Is there a chance this can happen with this change?

The only consequence is it's no longer possible to interrupt the process in the window between the test runner finishing up and the process normally ending after that. Or the transition window for `deno test --watch`.

If there was a real reason to overcome this upstream limitation, it's the effect on `Deno.removeSignalHandler("SIGINT")` and having compatible behaviour for Node's `process.on("SIGINT")` IMO.

> Also how will it interact if the code in tests registers a signal listener for SIGINT event?

The user's handler doesn't run because the process exits in our handler. I think that's fine, they should use subprocesses / integration tests for things like that.

**Comment by bartlomieju** [comment]
Thanks for detailed explanation, totally makes sense.

> If there was a real reason to overcome this upstream limitation, it's the effect on Deno.removeSignalHandler("SIGINT") and having compatible behaviour for Node's process.on("SIGINT") IMO.

Agreed, we should have one "custom" listener that just switches its actions once `Deno.removeSignalHandler("SIGINT")` is called.

File: cli/tools/test.rs [hunk 4]
```diff
@@ -153,13 +158,100 @@ pub enum TestOutput {
...
+      }
+      TestFailure::OverlapsWithSanitizers(long_names) => {
+        let mut string = "Started test step while another test step with sanitizers was running:".to_string();
+        for long_name in long_names {
+          string.push_str(&format!("\n  * {}", long_name));
+        }
+        string
+      }
+      TestFailure::HasSanitizersAndOverlaps(long_names) => {
+        let mut string = "Started test step with sanitizers while another test step was running:".to_string();
```
Comment (aapoalas) [inline cli/tools/test.rs:209]: **nitpick:** If I'm not badly mistaken, there's some way to join a vector of strings or possibly concatenate a string iterator, so it should be possible to do these error message creations without needing to use `push_str`. Not sure if it would be any better in any shape or form, though :)

**Review by aapoalas — APPROVED** [review]
LGTM. Left one absolutely useless nitpick about list of strings formatting.

**question:** Does the `color` crate automatically handle `NO_COLOR` flag, or should that be manually added to the added uses here? Current test reporter does respect the flag.

File: cli/tools/test.rs (hunk 4 above)
Comment (nayeemrmn) [reply cli/tools/test.rs:209]: I imagined that looking like `format!(..., foo.iter().map(|s| format!(..., s)).intersperse("\n").collect())` and went with this way. Would be nice if a macro existed like
```rs
concat2!(
  "Leaking async ops:\n",
  "  - {details[#]}\n",
  "To get more details where ops were leaked, run again with --trace-ops flag."
)
```

**Comment by nayeemrmn** [comment]
> **question:** Does the `color` crate automatically handle `NO_COLOR` flag, or should that be manually added to the added uses here? Current test reporter does respect the flag.

It's a local utility, yeah https://github.com/denoland/deno/blob/e4c60bc1c658be762f59d3c22b4040777a5100f8/runtime/colors.rs#L43-L45

**Comment by aapoalas** [comment]
@nayeemrmn Can we merge this?

**Comment by nayeemrmn** [comment]
Yeah no more changes to make
